Describe interaction between QUIC and TLS regarding saved 0-RTT state (#2947)
quicwg/base-drafts: draft-ietf-quic-tls.md

protection being called out specially.

~~~
+------------+                               +------------+
|            |<---- Handshake Messages ----->|            |
|            |<- Validate 0-RTT parameters ->|            |
|            |<--------- 0-RTT Keys ---------|            |
|            |<------- Handshake Keys -------|            |

> |   QUIC     |<--------- 1-RTT Keys ---------|    TLS     |

Reviewer: We probably need to move the QUIC and TLS labels up by a line.

## Interface to TLS

As shown in {{schematic}}, the interface from QUIC to TLS consists of four
primary functions:

- Sending and receiving handshake messages

> - Processing stored transport and application state from a 0-RTT capable session

Reviewer: s/0-RTT capable session ticket/resumed session/

```suggestion
- Processing stored transport and application state from a resumed session
```

As before with the mention of tickets; and early data is already covered.

>   ticket and determining if it is valid to accept early data on the connection

Reviewer: ```suggestion
  and determining if it is valid to accept early data
```

Negotiation packet.  These packets do not signify rejection of 0-RTT.

> ## Validating saved 0-RTT state

Reviewer: ```suggestion
## Validating 0-RTT Configuration
```

Author: Done.

When a server receives a ClientHello with the "early_data" extension, it has to

> decide whether to accept or rejct early data from the client. Some of this

Reviewer 2: ```suggestion
decide whether to accept or reject early data from the client. Some of this
```

> decision is made by the TLS stack (e.g. checking that the cipher suite being

Reviewer: ```suggestion
decision is made by the TLS stack (e.g., checking that the cipher suite being
```

Author: Done.

> resumed was included in the ClientHello, see Section 4.2.10 of {{!TLS13}}). QUIC

Reviewer 2: ```suggestion
resumed was included in the ClientHello; see Section 4.2.10 of {{!TLS13}}). QUIC
```

> saves additional transport and application state in a 0-RTT session ticket, and

Reviewer: QUIC doesn't save anything.  This is an implementation choice.  The key point is that QUIC requires that servers associate session configuration information with a resumed session.  The right thing to say here is that QUIC requires that a server be able to recover transport parameters for a resumed session, and different application protocols that use QUIC might have similar requirements regarding their configuration.  If transport parameters or application protocol configuration associated with a resumed session is not compatible with the current configuration of the server, the server MUST reject 0-RTT.

imposes additional restrictions on when a server may accept early data. The TLS
stack needs to check with the QUIC stack whether this saved state is valid for
accepting early data.

One example of such state that the QUIC stack checks when deciding whether or

> not to accept early data is the transport parameters extension. When HTTP/3

Reviewer: Transport parameters are the exhaustive set of things that the transport cares about.  But this implies otherwise.  I would concentrate on mentioning the h3 settings as an example of the sort of things that might be included.

Yes, we could define an extension to QUIC (the transport) that allowed state to be associated with a resumed sessions in a way that would affect this, but we don't need to do that now and our future selves can worry about how to integrate that with 0-RTT.

Author: I removed the mention of transport parameters specifically here, and rewrote this paragraph. My intention is to be vague so someone modifying a TLS stack to support QUIC doesn't say "let me check that transport params are compatible in the TLS stack, and then I'm done and don't need to provide a way for QUIC to reject early data", and instead actually delegates this decision to QUIC.

> ({{QUIC-HTTP}}) is used at the application layer, the SETTINGS frame from the

Reviewer 3: I think you want an informative reference here: `{{?QUIC-HTTP}}`

Author: It's already an informative ref, defined at the beginning of the file.

Reviewer 3: I was under the impression that we were changing to a mode where all references contain whether they're informative or normative.

Author: From doing git grep, the only reference to a QUIC-* draft that is marked with a `?` is the transport draft referencing QUIC-MANAGEABILITY. The rest omit a `?` or a `!`.

Reviewer 2: Because the QUIC-* drafts get published at the same time, the references between them are more manual than the others.  Therefore, they get explicitly listed as a normative/informative dependency and the inline status isn't needed.

previous session is another such example. This list is not intended to be

> exhaustive; future applications using QUIC may impose different restrictions.

Reviewer: ```suggestion
For example, HTTP/3 ({{QUIC-HTTP}}) settings determine how 0-RTT from the client
is interpreted. Other applications using QUIC could have different requirements
for determining whether to accept or reject 0-RTT.
```

Author: I incorporated similar language into the new/rewritten paragraph.

Negotiation packet.  These packets do not signify rejection of 0-RTT.

> ## Validating 0-RTT configuration

Reviewer: ```suggestion
## Validating 0-RTT Configuration
```

When a server receives a ClientHello with the "early_data" extension, it has to
decide whether to accept or reject early data from the client. Some of this
decision is made by the TLS stack (e.g., checking that the cipher suite being
resumed was included in the ClientHello; see Section 4.2.10 of {{!TLS13}}). Even
when the TLS stack has no reason to reject early data, the QUIC stack or the
application protocol using QUIC might reject early data because the
configuration of the transport or application associated with the resumed
session is not compatible with the server's current configuration.

QUIC requires additional transport state to be associated with a 0-RTT session
ticket. If stateless session tickets are used, this information must be stored

> in the session ticket. Application protocols that use QUIC might have similar

Reviewer 4: While it is true that additional transport state needs to be associated and verified when doing 0-RTT resumption, I am not sure if that necessarily mean that "this information must be stored in the session ticket."

For example, a server might store the associated states in a NEW_TOKEN token, that also contains the hash value of the session ticket. In such a design, when accepting a new connection, a server can consult the pair of the token and the session ticket, and decide if it can accept 0-RTT.

Reviewer 5: @nharper : Can you address this comment?

Author: Would it work to change "If stateless session tickets are used, this information must be stored in the session ticket." to "If stateless session tickets are used, this information may be stored in the session ticket."?

Reviewer: That would be fine.  The use of "must", implying "MUST", is a little too strong.

I would probably say that the choice of location (NEW_TOKEN vs NST) is an implementation-dependent one.  Implementations might want to use NST as that has better fate-sharing and they might not want to concern themselves with path validation for 0-RTT (or to allow a limited amount of 0-RTT in the case the token is bad).  Or, they could insist on a token and strong path validation, so moving the larger state necessary for 0-RTT into the token makes more sense.

Author: The intention of that sentence was to say that most likely you're using stateless tickets and storing this additional state in the ticket, so I reworded it to say that.

requirements regarding associating or storing state. This associated state is
used for deciding whether early data must be rejected. For example, HTTP/3
({{QUIC-HTTP}}) settings determine how early data from the client is

> interpreted. Other applications using QUIC could have different requiremenets

Reviewer: ```suggestion
interpreted. Other applications using QUIC could have different requirements
```

> for determining whether ot accept or reject early data.

Reviewer: ```suggestion
for determining whether to accept or reject early data.
```
